File: frontend-offline/src/ai/dangerMap.ts
import { ObjManager } from '../game/objManager'
import { TILE_X_NUM, TILE_Y_NUM, TILE_WIDTH } from '../game/gameConfig'

/**
 * Returns a 10x10 boolean grid where true means the tile is dangerous:
 * - Currently on fire
 * - Within blast range of an active bomb
 */
export function computeDangerMap(objManager: ObjManager): boolean[][] {
    const danger: boolean[][] = Array.from({ length: TILE_Y_NUM }, () =>
        Array(TILE_X_NUM).fill(false)
    )

    // Mark tiles currently on fire
    for (const fire of objManager.fires) {
        const allSprites = [fire.centerSprite, ...fire.verticalSprites, ...fire.horizontalSprites]
        for (const sprite of allSprites) {
            const ix = sprite.x / TILE_WIDTH
            const iy = sprite.y / TILE_WIDTH
            if (iy >= 0 && iy < TILE_Y_NUM && ix >= 0 && ix < TILE_X_NUM) {
                danger[iy][ix] = true
            }
        }
    }

    // Mark tiles in the blast range of active bombs
    const dirs: [number, number][] = [[0, -1], [0, 1], [-1, 0], [1, 0]]
    for (const bomb of objManager.bombs) {
        const { indexX, indexY } = bomb.getMapIndex()
        danger[indexY][indexX] = true

        for (const [dx, dy] of dirs) {
            for (let i = 1; i <= bomb.power; i++) {
                const nx = indexX + dx * i
                const ny = indexY + dy * i
                if (nx < 0 || nx >= TILE_X_NUM || ny < 0 || ny >= TILE_Y_NUM) break
                const tileType = objManager.mapManager.getMapTileTypeByIndex({ indexX: nx, indexY: ny })
                if (tileType === 'wall') break
                danger[ny][nx] = true
                if (tileType === 'brick' || tileType === 'bomb') break
            }
        }
    }

    return danger
}

File: py-game/ai/danger_map.py
from config import TILE_X_NUM, TILE_Y_NUM
from game.obj_manager import ObjManager


def compute_danger_map(obj_manager: ObjManager) -> list[list[bool]]:
    """
    Returns a TILE_Y_NUM × TILE_X_NUM boolean grid.
    True  → tile is currently on fire OR within blast range of an active bomb.
    """
    danger = [[False] * TILE_X_NUM for _ in range(TILE_Y_NUM)]

    # Tiles currently on fire
    for fire in obj_manager.fires:
        for ix, iy in fire.tiles:
            if 0 <= iy < TILE_Y_NUM and 0 <= ix < TILE_X_NUM:
                danger[iy][ix] = True

    # Blast radius of active bombs
    dirs = [(0, -1), (0, 1), (-1, 0), (1, 0)]
    for bomb in obj_manager.bombs:
        ix, iy = bomb.get_map_index()
        danger[iy][ix] = True

        for dx, dy in dirs:
            for i in range(1, bomb.power + 1):
                nx, ny = ix + dx * i, iy + dy * i
                if nx < 0 or nx >= TILE_X_NUM or ny < 0 or ny >= TILE_Y_NUM:
                    break
                tile_type = obj_manager.map_manager.get_tile_type(nx, ny)
                if tile_type == "wall":
                    break
                danger[ny][nx] = True
                if tile_type in ("brick", "bomb"):
                    break

    return danger
